### PR TITLE
Fix node url and http/https request types

### PR DESCRIPTION
**Pull request by jeffkenney**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
    - https://nodejs.org/dist/latest-v6.x/docs/api/url.html
    - https://nodejs.org/dist/latest-v6.x/docs/api/http.html#http_http_request_options_callback
    - https://nodejs.org/dist/latest-v6.x/docs/api/https.html#https_https_request_options_callback
- [] ~Increase the version number in the header if appropriate.~
- [] ~If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.~

### CODE REVIEW

File: types/node/v6/index.d.ts [hunk 1]
```diff
@@ -1346,7 +1346,7 @@ declare module "https" {
     };
     export interface Server extends tls.Server { }
     export function createServer(options: ServerOptions, requestListener?: Function): Server;
-    export function request(options: RequestOptions, callback?: (res: http.IncomingMessage) => void): http.ClientRequest;
+    export function request(options: RequestOptions | string, callback?: (res: http.IncomingMessage) => void): http.ClientRequest;
```
Comment (mhegazy) [inline types/node/v6/index.d.ts]: why is not this change made to other versions of node, i.e. v7, and v8?

**Comment by DanielRosenwasser** [comment]
Hey @jeffkenney could you apply the change to the other versions as well?

**Comment by jeffkenney** [comment]
@mhegazy @DanielRosenwasser I'll work on getting the other versions updated.

**Comment by bowdenk7** [comment]
Will merge in once the other version is updated. Thanks!

**Comment by jeffkenney** [comment]
updates applied to all versions, also backported the split Url / UrlObject types to v0 and v4

**Comment by andy-ms** [comment]
Hi @jeffkenney, could you fix the merge conflict? Then I'll take a quick look at the CI and hopefully merge. Thanks!

File: types/node/index.d.ts [hunk 2]
```diff
@@ -2156,37 +2156,29 @@ declare module "child_process" {
...
-        hash?: string;
+    export interface Url extends UrlObject {
+        port?: string;
+        query?: any;
     }
 
     export function parse(urlStr: string, parseQueryString?: boolean, slashesDenoteHost?: boolean): Url;
     export function format(URL: URL, options?: URLFormatOptions): string;
-    export function format(urlObject: UrlObject): string;
+    export function format(urlObject: UrlObject | string): string;
```
Comment (OliverJAsh) [inline types/node/index.d.ts]: Doesn't `format` use the same type as `parse` returns? If so, how come we need `UrlObject` and `Url`?

Also, I would suggest renaming `Url` to `ParsedURL` to disambiguate it from `URL`

Comment (Flarna) [reply types/node/index.d.ts]: From node docu: `The value of urlObject.port is coerced to a string and appended to result`
So in fact everything convertible to string is accepted here (which would be `any`) but you will get a `string | null` from `parse()`.
In real world apps `string | number` is usually what is needed which is most likely the reason why noone complained yet.
There are several such places in node; e.a. you can set everything convertible to `string` as a http header.

**Pull request merged**